src/internal/starlark/reflect.go

```diff
@@ -0,0 +1,240 @@
+		fv := rv.FieldByName(name)
+		if fv.Kind() == reflect.Invalid {
+			return nil, errors.Errorf("no field %q", name)
+		}
+		return Value(fv.Interface()), nil
+	}
+	return nil, errors.Errorf("cannot index into type %v (%v)", rv.Type(), rv.Kind())
+}
+func (r *Reflect) AttrNames() []string {
+	if r.Any == nil {
+		return nil
+	}
+	rv := reflect.ValueOf(r.Any)
+	if rv.Kind() == reflect.Ptr {
+		rv = rv.Elem()
+	}
+	if rv.Kind() == reflect.Struct {
+		var result []string
+		n := rv.NumField()
+		for i := 0; i < n; i++ {
+			result = append(result, rv.Type().Field(i).Name)
+		}
+		return result
+	}
+	return nil
+}
+func (r *Reflect) Cmp(y starlark.Value, _ int) (int, error) {
+	sr := FromStarlark(r)
+	sy := FromStarlark(y)
+	if reflect.DeepEqual(sr, sy) {
+		return 0, nil
+	}
+	return 0, errors.Errorf("cannot compare a %T (%v) to a %T (%v)", r, r, sy, sy)
+}
+
+// Value makes any Go value available to Starlark.
+func Value(in any) starlark.Value {
+	switch x := in.(type) {
+	case starlark.Value:
+		return x
+	case []starlark.Value:
+		return ReflectList(x)
+	case []byte:
+		return starlark.Bytes(string(x))
+	case string:
+		return starlark.String(x)
+	case bool:
+		return starlark.Bool(x)
+	case int:
+		return starlark.MakeInt(x)
+	case int8:
+		return starlark.MakeInt64(int64(x))
+	case int16:
+		return starlark.MakeInt64(int64(x))
+	case int32:
+		return starlark.MakeInt64(int64(x))
+	case int64:
+		return starlark.MakeInt64(x)
+	case uint:
+		return starlark.MakeUint(x)
+	case uint8:
+		return starlark.MakeUint64(uint64(x))
+	case uint16:
+		return starlark.MakeUint64(uint64(x))
+	case uint32:
+		return starlark.MakeUint64(uint64(x))
+	case uint64:
+		return starlark.MakeUint64(x)
+	case *big.Int:
+		return starlark.MakeBigInt(x)
+	case float32:
+		return starlark.Float(float64(x))
+	case float64:
+		return starlark.Float(x)
+	}
+	return value(reflect.ValueOf(in))
+}
+
+func value(v reflect.Value) starlark.Value {
+	if v.Kind() == reflect.Invalid {
+		return starlark.None
+	}
+	switch v.Kind() {
+	case reflect.Slice, reflect.Array:
+		n := v.Len()
+		var anys []any
+		for i := 0; i < n; i++ {
+			anys = append(anys, v.Index(i).Interface())
+		}
+		return ReflectList(anys)
+	case reflect.Map:
+		d := starlark.NewDict(v.Len())
+		iter := v.MapRange()
+		for iter.Next() {
+			kk := Value(iter.Key().Interface())
+			vv := Value(iter.Value().Interface())
+			if err := d.SetKey(kk, vv); err != nil {
+				// Unexpected; we can't handle it, so just do nothing.
+				return &Reflect{Any: v.Interface}
+			}
+		}
+		return d
+	}
+	return &Reflect{Any: v.Interface()}
+}
+
+// ReflectList returns a starlark.List from any Go slice.
+func ReflectList[T any](xs []T) *starlark.List {
+	var values []starlark.Value
+	for _, x := range xs {
+		values = append(values, Value(x))
+	}
+	return starlark.NewList(values)
+}
+
+// FromStarlark returns a Go value for a Starlark value.
+func FromStarlark(v starlark.Value) any {
+	switch x := v.(type) {
+	case *Reflect:
```

Comment:
when is v's type `*Reflect`?

Reply:
If you write a built-in function that returns a *Reflect and then another that takes a *Reflect, this is so you get the original Go value back out.  It's used in the TestRoundTrip for the SimpleStruct cases.